packages/client/wallets/smart-wallet/src/blockchain/wallets/SendTransactionService.ts

```diff
@@ -0,0 +1,137 @@
+import { SmartAccount } from "permissionless/accounts";
+import {
+    Abi,
+    Address,
+    BaseError,
+    Chain,
+    ContractFunctionRevertedError,
+    EstimateFeesPerGasReturnType,
+    Hex,
+    PublicClient,
+    TransactionReceipt,
+    Transport,
+    WaitForTransactionReceiptTimeoutError,
+} from "viem";
+
+import { CrossmintErrors, CrossmintSDKError } from "@crossmint/client-sdk-base";
+
+export type TransactionServiceTransactionRequest = {
+    address: Address;
+    abi: Abi;
+    functionName: string;
+    args: any;
+};
+
+export type TransactionServiceResendCallback = (resendAttempts: number) => Promise<{ resend: boolean }>;
+
+export class SendTransactionError extends CrossmintSDKError {
+    constructor(message: string, public readonly viemError: BaseError) {
+        super(message, CrossmintErrors.SEND_TRANSACTION_FAILED);
+    }
+}
+
+export class SendTransactionExecutionRevertedError extends CrossmintSDKError {
+    constructor(
+        message: string,
+        public readonly viemError: BaseError,
+        public readonly revertError: ContractFunctionRevertedError,
+        public txId: string | undefined
+    ) {
+        super(message, CrossmintErrors.SEND_TRANSACTION_FAILED);
+    }
+}
+
+export class SendTransactionService {
+    constructor(
+        private publicClient: PublicClient,
+        private gasIncreaseFactor = 1.2,
+        private confirmations = 2,
+        private transactionConfirmationTimeout = 15_000,
+        private defaultResendCallback: TransactionServiceResendCallback = async (resendAttempts) => ({
+            resend: resendAttempts <= 3,
+        })
+    ) {}
+
+    async sendTransaction(
+        request: TransactionServiceTransactionRequest,
+        client: SmartAccountClient<EntryPoint, Transport, Chain, SmartAccount<EntryPoint>>,
+        resendCallback = this.defaultResendCallback
+    ) {
+        for (let attempt = 0; attempt == 0 || (await resendCallback(attempt)).resend; attempt++) {
+            const gas = this.modifyGas(attempt, await this.publicClient.estimateFeesPerGas());
+            await this.simulateCall(request, undefined);
+            const hash = await client.writeContract({
+                ...request,
+                account: client.account,
+                chain: client.chain,
+                ...gas,
+            });
+            try {
+                const receipt = await this.publicClient.waitForTransactionReceipt({
+                    hash,
+                    confirmations: this.confirmations,
+                    timeout: this.transactionConfirmationTimeout,
+                });
+                return await this.handleReceipt(receipt, request);
+            } catch (e) {
+                if (e instanceof WaitForTransactionReceiptTimeoutError) {
+                    continue;
+                }
+                throw e;
+            }
+        }
+        // TODO discuss
+        throw new Error("Retries exceeded");
```

Comment:
I'm open to input on how we should handle this case, the most orthodox way would be to just export an error `SendTransactionRetriesExceededError extends CrossmintSDKError`